File: C#/NonBlockingFutureHolderSynchronizer/NonBlockingFutureHolderSynchronizer/TestNonBlockingFutureHolder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;


namespace NonBlockingFutureHolderSynchronizer
{
    class TestNonBlockingFutureHolder
    {

        private static bool GET_VALUE = true;
        private static bool SET_VALUE = false;

        public static void Main()
        {
            TestNonBlockingFutureHolder tests = new TestNonBlockingFutureHolder();
            tests.Test_simple_thread_waiting_for_integer_value() ;
            tests.Test_multiple_threads_waiting_for_integer_value();
            tests.Test_multiples_set_values_throw_IllegalStateException();
            tests.Test_simple_thread_terminate_by_zero_timeout() ;
            tests.Test_simple_thread_terminate_by_timeout();
            tests.Test_multiple_threads_but_one_terminate_by_timeout_and_the_other_get_the_value();
            Console.ReadLine();

        }

        public void Test_simple_thread_waiting_for_integer_value()
        {
 	        Console.WriteLine("---------simple_test_thread_waiting_for_integer_value() BEGIN------------");
            NonBlockingFutureHolder<String> futureHolder = new NonBlockingFutureHolder<String>();
            Thread t1 = new Thread(() => NonBlockingThread(futureHolder,1,null,2000,GET_VALUE));
            Thread t2 = new Thread(() => NonBlockingThread(futureHolder,2,"10",2000,SET_VALUE));
            

            t1.Start();
            Thread.Sleep(500);
            t2.Start();
            Thread.Sleep(500);
            Console.WriteLine("-------------------------------------------------------------------------");
        }

        public void Test_multiple_threads_waiting_for_integer_value()
        {
 	        Console.WriteLine("------test_multiple_threads_waiting_for_integer_value() BEGIN-------------");
            NonBlockingFutureHolder<string> futureHolder = new NonBlockingFutureHolder<string>();
            
            Thread t1 = new Thread(()=>NonBlockingThread(futureHolder,1,null,2000,GET_VALUE));
            Thread t2 = new Thread(()=>NonBlockingThread(futureHolder,2,null,2000,GET_VALUE));
            Thread t3 = new Thread(()=>NonBlockingThread(futureHolder,3,null,2000,GET_VALUE));
            Thread t4 = new Thread(()=>NonBlockingThread(futureHolder,4,"20",2000,SET_VALUE));

            t1.Start();
            t2.Start();
            t3.Start();
            Thread.Sleep(1000);
            t4.Start();
            Thread.Sleep(500);
        
            Console.WriteLine("-------------------------------------------------------------------------");
        }

        public void Test_multiples_set_values_throw_IllegalStateException()
        {
 	         Console.WriteLine("------test_multiples_set_values_throw_IllegalStateException() BEGIN------");
             NonBlockingFutureHolder<string> futureHolder = new NonBlockingFutureHolder<string>();

             Thread t1 = new Thread(() => NonBlockingThread(futureHolder, 1, "10",2000, SET_VALUE));
             Thread t2 = new Thread(() => NonBlockingThread(futureHolder,2,"20",2000,SET_VALUE));
            
            t1.Start();
            t2.Start();
            Thread.Sleep(200);
       
            Console.WriteLine("-------------------------------------------------------------------------");
        }

        public void Test_simple_thread_terminate_by_zero_timeout()
        {
 	         Console.WriteLine("--------test_simple_thread_terminate_by_zero_timeout() BEGIN--------------");
             NonBlockingFutureHolder<string> futureHolder = new NonBlockingFutureHolder<string>();
             
             Thread t1 = new Thread(() => NonBlockingThread(futureHolder,1,null,0,GET_VALUE));
             t1.Start();
             Thread.Sleep(10);
        
            Console.WriteLine("-------------------------------------------------------------------------");
        }

        public void Test_simple_thread_terminate_by_timeout()
        {
 	        Console.WriteLine("----------test_simple_thread_terminate_by_timeout() BEGIN----------------");
            NonBlockingFutureHolder<string> futureHolder = new NonBlockingFutureHolder<string>();
            
            Thread t1 = new Thread(() => NonBlockingThread(futureHolder,1,null,500,GET_VALUE));
            t1.Start();
            Thread.Sleep(2000);

            Console.WriteLine("--------------------------------------------------------------------------");
        }

        public void Test_multiple_threads_but_one_terminate_by_timeout_and_the_other_get_the_value()
        {
 	        Console.WriteLine("----test_multiple_threads_but_one_terminate_by_timeout_and_the_other_get_the_value() BEGIN---");
            NonBlockingFutureHolder<string> futureHolder = new NonBlockingFutureHolder<string>();
            Thread 
                t1 = new Thread(() =>NonBlockingThread(futureHolder,1,null,500,GET_VALUE)),
                t2 = new Thread(() =>NonBlockingThread(futureHolder,2,null,2000,GET_VALUE)),
                t3 = new Thread(() => NonBlockingThread(futureHolder,3,"10",1000,SET_VALUE));
            t1.Start();
            Thread.Sleep(2000);
            t2.Start();
            Thread.Sleep(500);
            t3.Start();
            Thread.Sleep(500);

            Console.WriteLine("----------------------------------------------------------------------------------------------");
        }   

        public void  NonBlockingThread(NonBlockingFutureHolder<string> nonBlockingFutureHolder, int id, string value, int timeout, bool isGetValue)
        {
             if (isGetValue) {
                Console.WriteLine("Thread with id {0} try get value", id);
                try
                {
                    value = nonBlockingFutureHolder.GetValue(timeout);
                }
                catch (TimeoutException e)
                {
                    Console.WriteLine(e.Message);
                    return;
                }
                    Console.WriteLine("thread with id {0} getValue = {1}",id,value );

            }
             else { 
                try {
                    Console.WriteLine("Thread with id {0} try to put value {1}", id,value);
                    nonBlockingFutureHolder.SetValue(value);
                    Console.WriteLine("Thread with id {0} put value {1} successfully",id,value);
                }catch(InvalidOperationException)
                {
                    Console.WriteLine("Thread with id {0} throw InvalidOperationException - Value already Inserted", id);
                }
                
                
            }
            }


   }

}

File: Java/NonBlockingFutureHolderSynchronizer/src/test/java/TestNonBlockingFutureHolder.java
import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * Created by Ruben Gomes on 20/06/2015.
 */
public class TestNonBlockingFutureHolder {

    private static final boolean GET_VALUE = true;
    private static final boolean SET_VALUE = false;


    public static class NonBlockingThread<T> extends Thread {

        NonBlockingFutureHolder<T> nonBlockingFutureHolder;
        int id;
        T value;
        long timeout;
        boolean isGetValue;


        public NonBlockingThread(NonBlockingFutureHolder<T> nonBlockingFutureHolder,int id, T value, long timeout, boolean isGetValue) {
            this.nonBlockingFutureHolder = nonBlockingFutureHolder;
            this.id = id;
            this.value = value;
            this.timeout = timeout;
            this.isGetValue = isGetValue;
        }

        @Override
        public void run() {
            if (isGetValue) {
                System.out.printf("Thread with id %d try get value \n", id);
                System.out.printf("thread with id %d getValue = %s \n",
                        id, (value = nonBlockingFutureHolder.getValue(timeout)) == null ? value+" by timeout" : value);

            }else
            try {
                System.out.printf("Thread with id %d try to put value %s \n", id,value);
                nonBlockingFutureHolder.setValue(value);
                System.out.printf("Thread with id %d put value %s successfully\n",id,value);
            }catch(IllegalStateException e){
                System.out.printf("Thread with id %d throw IllegalStateException - " + e.getMessage() + "\n", id);
            }
            }
    }

    @Test
    public void simple_test_thread_waiting_for_integer_value() throws InterruptedException {
        System.out.println("---------simple_test_thread_waiting_for_integer_value() BEGIN------------");
        NonBlockingFutureHolder<Integer> futureHolder = new NonBlockingFutureHolder<Integer>();
        NonBlockingThread<Integer>
                t1 = new NonBlockingThread<Integer>(futureHolder,1,null,5000,GET_VALUE),
                t2 = new NonBlockingThread<Integer>(futureHolder,2,10,5000,SET_VALUE);

        t1.start();
        Thread.sleep(200);
        t2.start();
        Thread.sleep(200);
        assertEquals(Integer.valueOf(10),futureHolder.getValue(5000));
        assertEquals(Thread.State.TERMINATED,t1.getState());
        assertEquals(Thread.State.TERMINATED,t2.getState());

        System.out.println("-------------------------------------------------------------------------");
    }


    @Test
    public void test_multiple_threads_waiting_for_integer_value() throws InterruptedException {
        System.out.println("------test_multiple_threads_waiting_for_integer_value() BEGIN-------------");
        NonBlockingFutureHolder<Integer> futureHolder = new NonBlockingFutureHolder<Integer>();
        NonBlockingThread<Integer>
                t1 = new NonBlockingThread<Integer>(futureHolder,1,null,5000,GET_VALUE),
                t2 = new NonBlockingThread<Integer>(futureHolder,2,null,5000,GET_VALUE),
                t3 = new NonBlockingThread<Integer>(futureHolder,3,null,5000,GET_VALUE),
                t4 = new NonBlockingThread<Integer>(futureHolder,4,20,5000,SET_VALUE);

        t1.start();
        t2.start();
        t3.start();
        Thread.sleep(500);
        t4.start();
        Thread.sleep(500);
        assertEquals(Thread.State.TERMINATED,t1.getState());
        assertEquals(Thread.State.TERMINATED,t2.getState());
        assertEquals(Thread.State.TERMINATED,t3.getState());
        assertEquals(Thread.State.TERMINATED,t4.getState());
        assertEquals(Integer.valueOf(20),futureHolder.getValue(5000));
        System.out.println("-------------------------------------------------------------------------");
    }

    @Test
    public void test_multiples_set_values_throw_IllegalStateException() throws InterruptedException{
        System.out.println("------test_multiples_set_values_throw_IllegalStateException() BEGIN------");
        NonBlockingFutureHolder<Integer> futureHolder = new NonBlockingFutureHolder<Integer>();
        NonBlockingThread<Integer>
                t1 = new NonBlockingThread<Integer>(futureHolder,1,10,5000,SET_VALUE),
                t2 = new NonBlockingThread<Integer>(futureHolder,2,20,5000,SET_VALUE);
        t1.start();
        Thread.sleep(200);
        assertEquals(Thread.State.TERMINATED, t1.getState());
        t2.start();
        Thread.sleep(200);
        assertEquals(Thread.State.TERMINATED,t2.getState());
        assertEquals(Integer.valueOf(10),futureHolder.getValue(5000));
        System.out.println("-------------------------------------------------------------------------");
    }

    @Test
    public void test_simple_thread_terminate_by_zero_timeout() throws InterruptedException{
        System.out.println("--------test_simple_thread_terminate_by_zero_timeout() BEGIN--------------");
        NonBlockingFutureHolder<Integer> futureHolder = new NonBlockingFutureHolder<Integer>();
        NonBlockingThread<Integer>
                t1 = new NonBlockingThread<Integer>(futureHolder,1,null,0,GET_VALUE);
        t1.start();
        Thread.sleep(10);
        assertEquals(Thread.State.TERMINATED, t1.getState());
        System.out.println("-------------------------------------------------------------------------");
    }

    @Test
    public void test_simple_thread_terminate_by_timeout() throws InterruptedException {
        System.out.println("----------test_simple_thread_terminate_by_timeout() BEGIN----------------");
        NonBlockingFutureHolder<Integer> futureHolder = new NonBlockingFutureHolder<Integer>();
        NonBlockingThread<Integer>
                t1 = new NonBlockingThread<Integer>(futureHolder,1,null,2000,GET_VALUE);
        t1.start();
        Thread.sleep(5000);
        assertEquals(Thread.State.TERMINATED, t1.getState());
        System.out.println("--------------------------------------------------------------------------");
    }



    @Test
    public void test_multiple_threads_but_one_terminate_by_timeout_and_the_other_get_the_value() throws InterruptedException {
        System.out.println("----test_multiple_threads_but_one_terminate_by_timeout_and_the_other_get_the_value() BEGIN---");
        NonBlockingFutureHolder<Integer> futureHolder = new NonBlockingFutureHolder<Integer>();
        NonBlockingThread<Integer>
                t1 = new NonBlockingThread<Integer>(futureHolder,1,null,2000,GET_VALUE),
                t2 = new NonBlockingThread<Integer>(futureHolder,2,null,4000,GET_VALUE),
                t3 = new NonBlockingThread<Integer>(futureHolder,3,10,1000,SET_VALUE);
        t1.start();
        t2.start();
        Thread.sleep(5000);
        assertEquals(Thread.State.TERMINATED, t1.getState());
        t3.start();
        Thread.sleep(500);
        assertEquals(Integer.valueOf(10),futureHolder.getValue(5000));
        assertEquals(Thread.State.TERMINATED, t1.getState());
        assertEquals(Thread.State.TERMINATED, t1.getState());
        assertEquals(Thread.State.TERMINATED, t1.getState());
        System.out.println("----------------------------------------------------------------------------------------------");

    }







}



